File: backend/app/schemas/live_game.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from datetime import datetime


class LiveGameParticipantResponse(BaseModel):
    """Response schema for live game participant"""
    summoner_name: str
    summoner_level: int
    champion_id: int
    spell1_id: int
    spell2_id: int
    team_id: int
    perks: Optional[Dict[str, Any]] = None
    
    class Config:
        from_attributes = True


class TeamCompositionResponse(BaseModel):
    """Response schema for team composition"""
    your_team: List[LiveGameParticipantResponse]
    enemy_team: List[LiveGameParticipantResponse]


class EnemyPlayerAnalysis(BaseModel):
    """Analysis of individual enemy player"""
    summoner_name: str
    champion_id: int
    summoner_level: int
    threat_level: str = Field(..., description="low, medium, or high")
    threat_reason: str
    counter_strategy: str
    estimated_rank: str = "Unknown"
    win_rate_estimate: Optional[float] = None
    main_role: str = "Unknown"


class TeamThreat(BaseModel):
    """Information about a team threat"""
    summoner_name: str
    champion_id: int
    reason: str
    counter_strategy: str


class TeamComposition(BaseModel):
    """Analysis of team composition"""
    damage_types: Dict[str, int] = {"physical": 0, "magical": 0, "true": 0}
    roles: Dict[str, int] = {"tank": 0, "damage": 0, "support": 0}
    crowd_control: str = "low"  # low, medium, high


class EnemyAnalysisResponse(BaseModel):
    """Response schema for enemy team analysis"""
    team_threats: List[TeamThreat]
    team_composition: TeamComposition
    individual_analysis: List[EnemyPlayerAnalysis]
    recommended_strategies: List[str]


class GameRecommendations(BaseModel):
    """Real-time game recommendations"""
    immediate_actions: List[str]
    item_builds: List[str]
    macro_strategy: List[str]
    warding_spots: List[str]


class LiveGameInfo(BaseModel):
    """Basic live game information"""
    game_id: str
    game_mode: Optional[str] = None
    game_type: Optional[str] = None
    map_id: Optional[int] = None
    queue_id: Optional[int] = None
    game_length: Optional[int] = None  # in seconds
    spectator_delay: Optional[int] = None


class LiveGameStatusResponse(BaseModel):
    """Complete live game status response"""
    is_in_game: bool
    game_info: Optional[LiveGameInfo] = None
    team_composition: Optional[TeamCompositionResponse] = None
    enemy_analysis: Optional[EnemyAnalysisResponse] = None
    recommendations: Optional[GameRecommendations] = None


class LiveGameMonitorResponse(BaseModel):
    """Response for monitoring multiple players"""
    success: bool
    data: Dict[str, Any]  # PUUID -> LiveGameStatusResponse
    region: str
    monitored_count: int


class FeaturedGameParticipant(BaseModel):
    """Participant in a featured game"""
    summoner_name: str
    champion_id: int
    team_id: int
    spell1_id: int
    spell2_id: int


class FeaturedGame(BaseModel):
    """Featured game from Riot API"""
    game_id: str
    game_mode: str
    game_length: int
    map_id: int
    participants: List[FeaturedGameParticipant]


class FeaturedGamesResponse(BaseModel):
    """Response for featured games"""
    game_list: List[FeaturedGame]
    client_refresh_interval: int


class LiveGameRecommendationsResponse(BaseModel):
    """Response for live game recommendations"""
    recommendations: GameRecommendations
    game_time_minutes: int
    game_phase: str = Field(..., description="early, mid, or late")
    next_major_objective: str


class CacheStatusResponse(BaseModel):
    """Cache status response"""
    total_entries: int
    size_estimate_mb: float
    oldest_entry_age_seconds: Optional[float] = None
    newest_entry_age_seconds: Optional[float] = None
    cleanup_task_running: bool


# Request schemas
class LiveGameAnalysisRequest(BaseModel):
    """Request for live game analysis"""
    puuid: str
    region: str = "na1"
    include_detailed_analysis: bool = True


class MonitorPlayersRequest(BaseModel):
    """Request to monitor multiple players"""
    puuids: List[str] = Field(..., max_items=10, description="List of PUUIDs to monitor")
    region: str = "na1"


# Error responses
class LiveGameErrorResponse(BaseModel):
    """Error response for live game endpoints"""
    success: bool = False
    error: str
    details: Optional[str] = None


# Champion counter information
class ChampionCounter(BaseModel):
    """Counter information for a specific champion"""
    champion_id: int
    counter_strategy: str
    difficulty: str = Field(..., description="easy, medium, or hard")
    key_abilities_to_avoid: List[str]
    recommended_items: List[str]


class LiveGameAnalyticsData(BaseModel):
    """Analytics data for live games"""
    total_games_detected: int
    average_game_length: float
    most_common_champions: List[Dict[str, Any]]
    win_rate_by_role: Dict[str, float]
    threat_level_distribution: Dict[str, int]


# Response wrappers
class LiveGameAPIResponse(BaseModel):
    """Generic API response wrapper"""
    success: bool
    data: Optional[Any] = None
    message: Optional[str] = None
    region: Optional[str] = None
    
    class Config:
        from_attributes = True


class SummonerLiveGameResponse(BaseModel):
    """Response including summoner info with live game status"""
    success: bool
    data: LiveGameStatusResponse
    region: str
    summoner: Dict[str, Any]  # Summoner info

File: sync-ui/src/types/live-game.ts
// Live Game Types for Frontend
export interface LiveGameParticipant {
  summoner_name: string;
  summoner_level: number;
  champion_id: number;
  spell1_id: number;
  spell2_id: number;
  team_id: number;
  perks?: {
    perkStyle?: number;
    perkSubStyle?: number;
    [key: string]: any;
  };
}

export interface TeamComposition {
  your_team: LiveGameParticipant[];
  enemy_team: LiveGameParticipant[];
}

export interface EnemyPlayerAnalysis {
  summoner_name: string;
  champion_id: number;
  summoner_level: number;
  threat_level: 'low' | 'medium' | 'high';
  threat_reason: string;
  counter_strategy: string;
  estimated_rank?: string;
  win_rate_estimate?: number;
  main_role?: string;
}

export interface TeamThreat {
  summoner_name: string;
  champion_id: number;
  reason: string;
  counter_strategy: string;
}

export interface TeamCompositionAnalysis {
  damage_types: {
    physical: number;
    magical: number;
    true: number;
  };
  roles: {
    tank: number;
    damage: number;
    support: number;
  };
  crowd_control: 'low' | 'medium' | 'high';
}

export interface EnemyAnalysis {
  team_threats: TeamThreat[];
  team_composition: TeamCompositionAnalysis;
  individual_analysis: EnemyPlayerAnalysis[];
  recommended_strategies: string[];
}

export interface GameRecommendations {
  immediate_actions: string[];
  item_builds: string[];
  macro_strategy: string[];
  warding_spots: string[];
}

export interface LiveGameInfo {
  game_id: string;
  game_mode?: string;
  game_type?: string;
  map_id?: number;
  queue_id?: number;
  game_length?: number; // in seconds
  spectator_delay?: number;
}

export interface LiveGameStatus {
  is_in_game: boolean;
  game_info?: LiveGameInfo;
  team_composition?: TeamComposition;
  enemy_analysis?: EnemyAnalysis;
  recommendations?: GameRecommendations;
}

export interface LiveGameResponse {
  success: boolean;
  data: LiveGameStatus;
  region: string;
  summoner: {
    puuid: string;
    summoner_name: string;
    region: string;
  };
}

export interface FeaturedGameParticipant {
  summoner_name: string;
  champion_id: number;
  team_id: number;
  spell1_id: number;
  spell2_id: number;
}

export interface FeaturedGame {
  game_id: string;
  game_mode: string;
  game_length: number;
  map_id: number;
  participants: FeaturedGameParticipant[];
}

export interface LiveGameRecommendationsResponse {
  recommendations: GameRecommendations;
  game_time_minutes: number;
  game_phase: 'early' | 'mid' | 'late';
  next_major_objective: string;
}

// Live game notification types
export interface LiveGameNotification {
  type: 'game_start' | 'game_end' | 'threat_detected' | 'objective_reminder';
  title: string;
  message: string;
  priority: 'low' | 'medium' | 'high';
  timestamp: Date;
  game_id?: string;
}

// Live game monitoring types
export interface PlayerMonitorStatus {
  puuid: string;
  summoner_name: string;
  is_in_game: boolean;
  game_status?: LiveGameStatus;
  last_checked: Date;
}

export interface MultiPlayerMonitorResponse {
  success: boolean;
  data: Record<string, LiveGameStatus>; // PUUID -> LiveGameStatus
  region: string;
  monitored_count: number;
}

// Champion counter types
export interface ChampionCounter {
  champion_id: number;
  counter_strategy: string;
  difficulty: 'easy' | 'medium' | 'hard';
  key_abilities_to_avoid: string[];
  recommended_items: string[];
}

// Threat level colors for UI
export const THREAT_LEVEL_COLORS = {
  low: 'text-green-400',
  medium: 'text-yellow-400',
  high: 'text-red-400'
} as const;

export const THREAT_LEVEL_BACKGROUNDS = {
  low: 'bg-green-900/20 border-green-500/30',
  medium: 'bg-yellow-900/20 border-yellow-500/30',
  high: 'bg-red-900/20 border-red-500/30'
} as const;

// Game phase colors
export const GAME_PHASE_COLORS = {
  early: 'text-blue-400',
  mid: 'text-yellow-400',
  late: 'text-red-400'
} as const;

// Team ID constants
export const TEAM_IDS = {
  BLUE: 100,
  RED: 200
} as const;
